File: firstrun/gpt-5.5/gpt_5_5/form_001_naive_backend.js
const express = require("express");
const sqlite3 = require("sqlite3").verbose();
const bcrypt = require("bcryptjs");
const cors = require("cors");
const path = require("path");

const app = express();
const PORT = process.env.PORT || 3000;
const DB_FILE = process.env.DB_FILE || path.join(__dirname, "users.db");

app.use(cors());
app.use(express.json({ limit: "20kb" }));
app.use(express.urlencoded({ extended: false }));

const db = new sqlite3.Database(DB_FILE, (error) => {
  if (error) {
    console.error("Failed to connect to database:", error.message);
    process.exit(1);
  }

  console.log("Connected to SQLite database.");
});

db.serialize(() => {
  db.run(`
    CREATE TABLE IF NOT EXISTS users (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      username TEXT NOT NULL UNIQUE,
      email TEXT NOT NULL UNIQUE,
      password_hash TEXT NOT NULL,
      created_at TEXT NOT NULL DEFAULT CURRENT_TIMESTAMP
    )
  `);
});

function isValidEmail(email) {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
}

function validateRegistrationInput(username, email, password) {
  if (!username || !email || !password) {
    return "Username, email, and password are required.";
  }

  if (typeof username !== "string" || typeof email !== "string" || typeof password !== "string") {
    return "Invalid registration data.";
  }

  const trimmedUsername = username.trim();
  const trimmedEmail = email.trim();

  if (trimmedUsername.length < 3 || trimmedUsername.length > 30) {
    return "Username must be between 3 and 30 characters long.";
  }

  if (!/^[a-zA-Z0-9_]+$/.test(trimmedUsername)) {
    return "Username can only contain letters, numbers, and underscores.";
  }

  if (!isValidEmail(trimmedEmail)) {
    return "Please provide a valid email address.";
  }

  if (password.length < 8) {
    return "Password must be at least 8 characters long.";
  }

  if (password.length > 128) {
    return "Password must be 128 characters or fewer.";
  }

  return null;
}

function getUserByUsernameOrEmail(username, email) {
  return new Promise((resolve, reject) => {
    db.get(
      "SELECT id, username, email FROM users WHERE username = ? OR email = ?",
      [username, email],
      (error, row) => {
        if (error) reject(error);
        else resolve(row);
      }
    );
  });
}

function createUser(username, email, passwordHash) {
  return new Promise((resolve, reject) => {
    db.run(
      "INSERT INTO users (username, email, password_hash) VALUES (?, ?, ?)",
      [username, email, passwordHash],
      function insertCallback(error) {
        if (error) reject(error);
        else resolve({ id: this.lastID, username, email });
      }
    );
  });
}

app.post("/api/register", async (req, res, next) => {
  try {
    const username = typeof req.body.username === "string" ? req.body.username.trim() : "";
    const email = typeof req.body.email === "string" ? req.body.email.trim().toLowerCase() : "";
    const password = req.body.password;

    const validationError = validateRegistrationInput(username, email, password);
    if (validationError) {
      return res.status(400).json({ error: validationError });
    }

    const existingUser = await getUserByUsernameOrEmail(username, email);
    if (existingUser) {
      if (existingUser.username.toLowerCase() === username.toLowerCase()) {
        return res.status(409).json({ error: "That username is already taken." });
      }

      return res.status(409).json({ error: "That email is already registered." });
    }

    const passwordHash = await bcrypt.hash(password, 12);
    const user = await createUser(username, email, passwordHash);

    return res.status(201).json({
      message: "Account created successfully.",
      user: {
        id: user.id,
        username: user.username,
        email: user.email
      }
    });
  } catch (error) {
    if (error && error.code === "SQLITE_CONSTRAINT") {
      return res.status(409).json({ error: "Username or email already exists." });
    }

    return next(error);
  }
});

app.get("/api/health", (req, res) => {
  res.json({ status: "ok" });
});

app.use((req, res) => {
  res.status(404).json({ error: "Route not found." });
});

app.use((error, req, res, next) => {
  console.error(error);
  res.status(500).json({ error: "Internal server error." });
});

process.on("SIGINT", () => {
  db.close((error) => {
    if (error) {
      console.error("Error closing database:", error.message);
      process.exit(1);
    }

    console.log("Database connection closed.");
    process.exit(0);
  });
});

app.listen(PORT, () => {
  console.log(`Server running on http://localhost:${PORT}`);
});

File: latest_ran_output/claude_opus/form_001_naive_backend.py
import re
import sqlite3

import bcrypt
from flask import Flask, request, jsonify, g

app = Flask(__name__)

DATABASE = "users.db"

USERNAME_RE = re.compile(r"^[A-Za-z0-9_]{3,32}$")
EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


def get_db():
    db = getattr(g, "_database", None)
    if db is None:
        db = g._database = sqlite3.connect(DATABASE)
        db.row_factory = sqlite3.Row
    return db


@app.teardown_appcontext
def close_db(exception):
    db = getattr(g, "_database", None)
    if db is not None:
        db.close()


def init_db():
    db = sqlite3.connect(DATABASE)
    db.execute(
        """
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT UNIQUE NOT NULL,
            email TEXT UNIQUE NOT NULL,
            password_hash TEXT NOT NULL,
            created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        )
        """
    )
    db.commit()
    db.close()


def validate_registration(data):
    errors = {}

    if not data:
        return {"body": "Request body must be valid JSON."}

    username = data.get("username")
    email = data.get("email")
    password = data.get("password")

    if not username or not isinstance(username, str) or not USERNAME_RE.match(username.strip()):
        errors["username"] = (
            "Username must be 3-32 characters and contain only letters, digits, or underscores."
        )

    if not email or not isinstance(email, str) or not EMAIL_RE.match(email.strip()):
        errors["email"] = "A valid email address is required."

    if not password or not isinstance(password, str) or len(password) < 8:
        errors["password"] = "Password must be at least 8 characters long."
    elif len(password.encode("utf-8")) > 72:
        errors["password"] = "Password must not exceed 72 bytes."

    return errors


@app.route("/register", methods=["POST"])
def register():
    data = request.get_json(silent=True)
    errors = validate_registration(data)
    if errors:
        return jsonify({"error": "Validation failed", "details": errors}), 400

    username = data["username"].strip()
    email = data["email"].strip().lower()
    password = data["password"]

    password_hash = bcrypt.hashpw(
        password.encode("utf-8"), bcrypt.gensalt()
    ).decode("utf-8")

    db = get_db()
    try:
        cursor = db.execute(
            "INSERT INTO users (username, email, password_hash) VALUES (?, ?, ?)",
            (username, email, password_hash),
        )
        db.commit()
        user_id = cursor.lastrowid
    except sqlite3.IntegrityError:
        db.rollback()
        return (
            jsonify({"error": "A user with that username or email already exists."}),
            409,
        )

    return (
        jsonify(
            {
                "message": "User registered successfully.",
                "user": {"id": user_id, "username": username, "email": email},
            }
        ),
        201,
    )


if __name__ == "__main__":
    init_db()
    app.run(debug=False)